sqlmail.go

```diff
@@ -0,0 +1,168 @@
+package maddy
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/emersion/go-imap/backend"
+	"github.com/emersion/maddy/config"
+	"github.com/emersion/maddy/log"
+	"github.com/emersion/maddy/module"
+	"github.com/foxcpp/go-sqlmail"
+	imapsqlmail "github.com/foxcpp/go-sqlmail/imap"
+)
+
+type SQLMail struct {
+	*imapsqlmail.Backend
+	instName string
+	Log      log.Logger
+}
+
+type Literal struct {
+	io.Reader
+	length int
+}
+
+func (l Literal) Len() int {
+	return l.length
+}
+
+func (sqlm *SQLMail) Name() string {
+	return "sqlmail"
+}
+
+func (sqlm *SQLMail) InstanceName() string {
+	return sqlm.instName
+}
+
+func (sqlm *SQLMail) Version() string {
+	return sqlmail.VersionStr
+}
+
+func NewSQLMail(_, instName string) (module.Module, error) {
+	return &SQLMail{
+		instName: instName,
+		Log:      log.Logger{Out: log.StderrLog, Name: "sqlmail"},
+	}, nil
+}
+
+func (sqlm *SQLMail) Init(globalCfg map[string]config.Node, rawCfg config.Node) error {
+	var driver string
+	var dsn string
+	appendlimitVal := int64(-1)
+
+	opts := imapsqlmail.Opts{}
+	cfg := config.Map{}
+	cfg.String("driver", false, true, "", &driver)
+	cfg.String("dsn", false, true, "", &dsn)
+	cfg.Int64("appendlimit", false, false, 32*1024*1024, &appendlimitVal)
+	cfg.Bool("debug", true, &sqlm.Log.Debug)
+
+	if _, err := cfg.Process(globalCfg, &rawCfg); err != nil {
+		return err
+	}
+
+	if appendlimitVal == -1 {
+		opts.MaxMsgBytes = nil
+	} else {
+		opts.MaxMsgBytes = new(uint32)
+		*opts.MaxMsgBytes = uint32(appendlimitVal)
+	}
+	back, err := imapsqlmail.NewBackend(driver, dsn, opts)
+	if err != nil {
+		return fmt.Errorf("sqlmail: %s", err)
+	}
+	sqlm.Backend = back
+
+	return nil
+}
+
+func (sqlm *SQLMail) Extensions() []string {
```

Comment:
Maybe we should rename this to `IMAPExtensions` -- we might have things like `SMTPExtensions` in the future.

Reply:
Makes sense, will rename.